Ceph object store for external cluster

aruniiird opened the pull request:
CephObjectStore creation in external cluster mode according to the latest rook-ceph changes

aruniiird commented:
@leseb , can you please take a look.

leseb commented on pkg/controller/storagecluster/initialization_reconciler.go:
```diff
@@ -374,6 +384,50 @@ func (r *ReconcileStorageCluster) ensureCephObjectStores(instance *ocsv1.Storage
+	initData *ocsv1.StorageCluster, rgwEndpoint string, reqLogger logr.Logger) ([]*cephv1.CephObjectStore, error) {
+	rgwEndpoint = strings.TrimSpace(rgwEndpoint)
+	var gateWay cephv1.GatewaySpec
+	if rgwEndpoint != "" {
```
I don't think it can be empty, the UI will not accept that, or perhaps are you trying to catch a manual deployment issue?

aruniiird replied:
@leseb , there was discussion regarding making the `rgw-endpoint` optional and this function should serve both the needs.

leseb commented on pkg/controller/storagecluster/initialization_reconciler.go:
```diff
@@ -205,9 +208,14 @@ func (r *ReconcileStorageCluster) ensureExternalStorageClusterResources(instance
+				externalRgwEndpointReplaceColon := strings.Replace(rgwEndpoint, ":", "_", -1)
 				externalRgwEndpoint = externalRgwEndpointReplaceColon
 
 				// Setting the Endpoint for OBC StorageClass
```
We don't need to set the endpoint anymore on the SC for OBC. We just need to set the `objectStoreName` back to it, just like we do for the converged case.

aruniiird replied:
Made the changes, @leseb , please take a look...

aruniiird commented:
Rebased this PR on top of @umangachapagain 's [Update rook to 1.4](https://github.com/openshift/ocs-operator/pull/657) PR.
Removed all the `TODO` s to working code. @leseb , please take a look...

aruniiird commented:
This PR depends on https://github.com/openshift/ocs-operator/pull/657

aruniiird commented:
@anmolsachan , this is the PR for the new CephObjectStore changes. Please take a look...

leseb commented on pkg/controller/storagecluster/initialization_reconciler.go:
```diff
@@ -374,6 +385,59 @@ func (r *ReconcileStorageCluster) ensureCephObjectStores(instance *ocsv1.Storage
+	}
+	if hostIP == "" {
+		err := fmt.Errorf("An empty rgw host 'IP' address found")
+		reqLogger.Error(err, err.Error())
```
Why not just `reqLogger.Error(err)`?

aruniiird replied:
`reqLogger.Error` requires a second string argument, which provides a message. Will change to a different message (rather than the same error message).

leseb commented on pkg/controller/storagecluster/initialization_reconciler.go:
```diff
@@ -374,6 +385,59 @@ func (r *ReconcileStorageCluster) ensureCephObjectStores(instance *ocsv1.Storage
+	initData *ocsv1.StorageCluster, rgwEndpoint string, reqLogger logr.Logger) ([]*cephv1.CephObjectStore, error) {
+	// check whether the provided rgw endpoint is empty
+	if rgwEndpoint = strings.TrimSpace(rgwEndpoint); rgwEndpoint == "" {
+		reqLogger.Info("WARNING: Empty RGW Endpoint specified, external CephObjectStore won't be created")
```
Is it a warning or an error? Either way, use the appropriate `reqLogger`.

aruniiird replied:
Yes it is a `Warning` only and, in a normal scenario, this function won't be even called (as won't be propagating the `ceph-rgw` in the json output).
`reqLogger` (an instance of `github.com/go-logr/logr` interface) don't have a `.Warn` (and have a [philosophy](https://dave.cheney.net/2015/11/05/lets-talk-about-logging) on why no warnings). So `Info` is the best (and non-complicated) bet to go.

leseb commented on pkg/controller/storagecluster/initialization_reconciler.go:
```diff
@@ -374,6 +385,59 @@ func (r *ReconcileStorageCluster) ensureCephObjectStores(instance *ocsv1.Storage
+		return nil, nil
+	}
+
+	var gateWay cephv1.GatewaySpec
```
What about a method that returns a `cephv1.GatewaySpec` instead of this long method? It will be easier to unit test too.

aruniiird replied:
Done...

leseb commented on pkg/controller/storagecluster/initialization_reconciler.go:
```diff
@@ -145,6 +147,7 @@ func (r *ReconcileStorageCluster) ensureExternalStorageClusterResources(instance
 		Kind:       instance.Kind,
 		Name:       instance.Name,
 	}
+	var extCephObjectStores []*cephv1.CephObjectStore
```
Why an array?

aruniiird replied:
This is to pass it to `createCephObjectStoreInstances` method (which takes an array of `*cephv1.CephObjectStore` instances as first argument).

leseb commented on pkg/controller/storagecluster/generate.go:
```diff
@@ -27,7 +27,12 @@ func generateNameForCephBlockPool(initData *ocsv1.StorageCluster) string {
 }
 
 func generateNameForCephObjectStore(initData *ocsv1.StorageCluster) string {
```
Do you have unit tests for this?

aruniiird replied:
Not yet, will add one...

aruniiird replied:
Added unit tests...

leseb commented on pkg/controller/storagecluster/initialization_reconciler.go:
```diff
@@ -205,24 +208,36 @@ func (r *ReconcileStorageCluster) ensureExternalStorageClusterResources(instance
 				// Set the external rgw endpoint variable for later use on the Noobaa CR (as a label)
 				// Replace the colon with an underscore, otherwise the label will be invalid
-				externalRgwEndpointReplaceColon := strings.Replace(d.Data[externalCephRgwEndpointKey], ":", "_", -1)
+				externalRgwEndpointReplaceColon := strings.Replace(rgwEndpoint, ":", "_", -1)
```
I need to check with @nb-ohad, for changes in Noobaa but this might need to go away as well.

jarrpa commented:
/ok-to-test

leseb commented:
@aruniiird please rebase

jarrpa commented on pkg/controller/storagecluster/cephobjectstores.go:
```diff
@@ -26,16 +26,14 @@ func (r *ReconcileStorageCluster) ensureCephObjectStores(instance *ocsv1.Storage
+		err = r.createCephObjectStores(cephObjectStores, instance, reqLogger)
+	}
 	if err != nil {
+		reqLogger.Error(err, "could not create CephObjectStores")
```
These changes don't seem related to the commit message. This can probably be integrated into the next commit.

aruniiird replied:
Moved to a separate commit...

jarrpa commented on pkg/controller/storagecluster/cephobjectstores.go:
```diff
@@ -28,7 +28,7 @@ func (r *ReconcileStorageCluster) ensureCephObjectStores(instance *ocsv1.Storage
 	var cephObjectStores []*cephv1.CephObjectStore
 	if cephObjectStores, err = r.newCephObjectStoreInstances(instance); err == nil {
-		err = r.createCephObjectStores(cephObjectStores, instance, reqLogger)
+		err = r.createCephObjectStores(cephObjectStores, reqLogger)
```
Why is this being removed here? Was it never used?

aruniiird replied:
It was never used in the `createCephObjectStores()` function.

jarrpa commented on pkg/controller/storagecluster/cephobjectstores.go:
```diff
@@ -38,28 +38,35 @@ func (r *ReconcileStorageCluster) ensureCephObjectStores(instance *ocsv1.Storage
 			if existing.DeletionTimestamp != nil {
-				reqLogger.Info(fmt.Sprintf("Unable to restore init object because %s is marked for deletion", existing.Name))
-				return fmt.Errorf("failed to restore initialization object %s because it is marked for deletion", existing.Name)
+				err := fmt.Errorf("failed to restore initialization object %s because it is marked for deletion",
+					existing.Name)
+				reqLogger.Error(err,
+					fmt.Sprintf("Unable to restore init object because %s is marked for deletion", existing.Name))
+				return err
```
Why are there so many error message changes? These should be in a separate commit. Also, please don't wrap the fmt/logging function lines.

aruniiird replied:
Corrected the error message and moved to the refactoring commit...

jarrpa commented on pkg/controller/storagecluster/external_resources.go:
```diff
@@ -141,9 +205,17 @@ func (r *ReconcileStorageCluster) createExternalStorageClusterResources(scs []*s
+				}
+				// rgw-endpoint is no longer needed in the 'd.Data' dictionary,
+				// and can be deleted
+				delete(d.Data, externalCephRgwEndpointKey)
```
What's the point of doing this? What happens if you don't delete it?

aruniiird replied:
RGW endpoint details we already stored in a `CephObjectStore`. If you don't `delete()` , the rgwEndpoint key / value will be again added to the StorageClass' `Parameters` map, which is not needed.

In the previous design, we used to store the rgw-endpoint details in the StorageClass itself, but now rgw-endpoint details are stored in a `CephObjectStore` object and in the SC, as one of the parameter, we just point to this cephobjectstore's name.

jarrpa replied:
...so then why is this still being passed in the RGW StorageClass data? There should be a separate place for it so that this hack isn't necessary.

aruniiird replied:
What I understood was,
We didn't make any change in the rook's JSON external input creation script (and kept it backward compatible). Thus changes were done only on OCS-Operator side with the same receiving input file.

jarrpa replied:
Thank you!!

anmolsachan commented on pkg/controller/storagecluster/external_resources.go:
```diff
@@ -77,15 +76,86 @@ func (r *ReconcileStorageCluster) ensureExternalStorageClusterResources(instance
+		return nil, err
+	}
+	gateWay.ExternalRgwEndpoints = []corev1.EndpointAddress{{IP: hostIP}}
+	if portInt64, err := strconv.ParseInt(portStr, 10, 32); err != nil {
```
```suggestion
	portInt64, err := strconv.ParseInt(portStr, 10, 32)
	if err != nil {
		reqLogger.Error(err,
			fmt.Sprintf("invalid rgw 'port' provided: %s", portStr))
		return nil, err
	}
	gateWay.Port = int32(portInt64)
```

anmolsachan replied:
Can you please refractor it a little? It's giving lint error

aruniiird replied:
Corrected...

aruniiird commented:
Added test cases... Ready for final reviews ... @jarrpa , @leseb please see

jarrpa commented on pkg/controller/storagecluster/external_resources.go:
```diff
@@ -205,16 +282,20 @@ func (r *ReconcileStorageCluster) createExternalStorageClusterConfigMap(cm *core
 			reqLogger.Error(err, "unable the get the configmap")
-			return err
 		}
+		return err
```
Don't do this, leave the return statements as-is.

aruniiird replied:
Yup, changed it ...

jarrpa commented on pkg/controller/storagecluster/external_resources.go:
```diff
@@ -205,16 +282,20 @@ func (r *ReconcileStorageCluster) createExternalStorageClusterConfigMap(cm *core
 	return nil
 }
 
+/*
+go: found golang.org/x/lint/golint in golang.org/x/lint v0.0.0-20200302205851-738671d3881b
+./pkg/controller/storagecluster/external_resources.go:109:9: if block ends with a return statement, so drop this else and outdent its block (move short variable declaration to its own line if necessary)
+*/
+
```
Is this supposed to be here?

aruniiird replied:
Removed it...

jarrpa commented on pkg/controller/storagecluster/storageclasses.go:
```diff
@@ -84,14 +84,9 @@ func (r *ReconcileStorageCluster) newOBCStorageClass(initData *ocsv1.StorageClus
 		Parameters: map[string]string{
 			"objectStoreNamespace": initData.Namespace,
 			"region":               "us-east-1",
+			"objectStoreName":      generateNameForCephObjectStore(initData),
```
So now we're generating this for both external and converged modes? Did something that drastic really change in the OBC StorageClass?

aruniiird replied:
A minor change in the OBC StorageClass' `Parameter map[string]string` now takes `objectStoreName`, instead of direct `endpoint`.

jarrpa replied:
I'd hardly call this minor. :) Also, you didn't answer my question: this is for both external and converged modes? How does it get used in converged mode, and how did it work before without this?

aruniiird replied:
Oops! missed the question mark (read the first line as a statement not as question)...
Yes, we are generating the CephObjectStore for both external and converged mode... But the creation of `CephObjectStore` is different in each mode,
a. Converged mode - we use ensure function to create CephObjectStore
b. External mode - we create it when `ceph-rgw` storageclass is available with the endpoint data in it.

@leseb ^^^

jarrpa commented on pkg/controller/storagecluster/cephobjectstores.go:
```diff
@@ -55,12 +55,16 @@ func (r *ReconcileStorageCluster) createCephObjectStores(cephObjectStores []*cep
 			err = r.client.Create(context.TODO(), cephObjectStore)
 			if err != nil {
+				reqLogger.Error(err,
+					fmt.Sprintf("Creation failed for CephObjectStore object: %s", cephObjectStore.Name))
```
Move all changes to this file in this second commit to the first commit. These are part of the refactoring and do not impact the addition of the external CephObjectStore.

aruniiird replied:
Done...

jarrpa commented:
Holding until #657 merges.

/hold

aruniiird commented on pkg/controller/storagecluster/cephobjectstores.go:
```diff
@@ -26,16 +26,14 @@ func (r *ReconcileStorageCluster) ensureCephObjectStores(instance *ocsv1.Storage
-	if err != nil {
-                reqLogger.Error(err,"could not create CephObjectStores")
-                return err
-        }
```
Done...

aruniiird commented on pkg/controller/storagecluster/cephobjectstores.go:
```diff
@@ -47,8 +45,9 @@ func (r *ReconcileStorageCluster) createCephObjectStores(cephObjectStores []*cep
-				reqLogger.Info(fmt.Sprintf("Unable to restore init object because %s is marked for deletion", existing.Name))
-				return fmt.Errorf("failed to restore initialization object %s because it is marked for deletion", existing.Name)
+				err := fmt.Errorf("failed to restore cephobjectstore object %s because it is marked for deletion", existing.Name)
+				reqLogger.Error(err, "cephobjectstore restore failed")
```
Done...

aruniiird commented:
@jarrpa , addressed all the review comments... Please take a look...

jarrpa commented:
/hold cancel
